openvdb/points/PointMask.h

```diff
@@ -304,18 +295,13 @@ inline typename GridT::Ptr convertPointsToScalar(
 
     tree::LeafManager<const typename PointDataGridT::TreeType> leafManager(points.tree());
 
-    const bool useGroup = !includeGroups.empty() || !excludeGroups.empty();
-
-    if (useGroup) {
-        const auto leaf = points.constTree().cbeginLeaf();
-        MultiGroupFilter filter(includeGroups, excludeGroups, leaf->attributeSet());
-        PointsToTransformedScalarOp<GridT, PointDataGridT, MultiGroupFilter> pointsToScalarOp(
-            transform, pointsTransform, filter, combiner);
-        tbb::parallel_for(leafManager.leafRange(), pointsToScalarOp);
-    }
-    else {
-        NullFilter filter;
+    if (filter.state() == index::ALL) {
```

Comment:
Is this static state method really needed? It seems redundant as the only filter capable of returning ALL is a null filter anyway? I would have expected callers of function requiring a filter to perform some filtering and use a NullFilter otherwise (which should be the default argument in the majority of cases). Imo that method can be removed, simplifying the filter object and branching patterns within all methods.

Reply:
Actually, the MultiGroupFilter can also return ALL. One of the key ideas here is to eliminate some of the complexity for the caller - in adopting this the caller no longer needs to check their own include / exclude arguments and determine whether to branch their code to use a NullFilter or a MultiGroupFilter, they can simply call with a MultiGroupFilter and default empty include / exclude arguments and the underlying logic will branch accordingly to pick up the faster implementation.

Reply:
I don't agree it would be clearer - from the user perspective, the API is simpler because they can use MultiGroupFilter and not worry about the filtering performance. You're right that it trades a little compile-time cost and a little extra IndexFilter complexity, but in my opinion I think that's worthwhile.

There is a filter.state() == index::NONE check, but this is in the creation of the iterator. Not as useful, but I included it for consistency. Also, regardless of implementation, there should always be an input to the MultiGroupFilter that evaluates all points, so whatever that input is should see this branching optimization (in reference to your last comment).

Reply:
They can use MultiGroupFilter and not worry about filtering performance: would this not also be true if the state function that took a leaf was implemented (perhaps a very minor penalty but likely negligible). I thought this is what this method was for and was my main point in my previous comment.

trades a little compile-time cost: didn't mention this, don't think this is a factor

I think this confuses things slightly and not sure I see its worth but oh well!